languages/tree-sitter-stack-graphs-typescript/build.rs

```diff
@@ -0,0 +1,98 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2024, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+use std::collections::HashSet;
+use std::path::Path;
+
+use anyhow::{Result, anyhow, bail};
+use regex::Regex;
+
+/// A stack of dialects as selected by the directives
+#[derive(Debug, Default)]
+struct DialectStack(Vec<HashSet<String>>);
+
+impl DialectStack {
+    fn push(&mut self, values: Vec<String>) -> Result<()> {
+        // ensure that the new values are a subset of the current scope
+        if let Some(current) = self.0.last() {
+            if !values.iter().all(|v| current.contains(v)) {
+                bail!("Directive values are not a subset of the current scope");
+            }
+        }
+
+        self.0.push(values.into_iter().collect());
+
+        Ok(())
+    }
+
+    fn pop(&mut self) -> Result<()> {
+        if let Some(_) = self.0.pop() {
+            Ok(())
+        } else {
+            Err(anyhow!("Directive stack is empty"))
+        }
+    }
+
+    fn contains(&self, query: &str) -> bool {
+        if let Some(current) = self.0.last() {
+            current.contains(query)
+        } else {
+           true
+        }
+    }
+}
+
+/// preprocess the input file, removing lines that are not for the selected dialect
+fn preprocess(input: impl std::io::Read, mut output: impl std::io::Write, dialect: &str) -> anyhow::Result<()> {
+    // Matches:   ; # dialect typescript tsx
+    let directive_start = Regex::new(r";[ \t]*#[ \t]*dialect[ \t]+([a-zA-Z\t ]+)").unwrap();
```

Comment:
A few nits:

- use predefined classes like `\w` and `\s` to make it a little more compact,
- treat `#dialect` as one thing, don't allow whitespace between the has and the keyword.

```suggestion
    let directive_start = Regex::new(r";\s*#dialect\s+(\w+)").unwrap();
```

(Note that this includes the change that we expect a single dialect name, not a list.)

Reply:
That's a good catch. I looked in the [syntax section](https://docs.rs/regex/latest/regex/#syntax) and didn't see `\w` and friends and was very surprised. But it looks they're just documented elsewhere.